darwin/torch/dataset.py

```diff
@@ -0,0 +1,276 @@
+
+from darwin.client import Client
+from darwin.torch.utils import (
+    convert_polygon_to_sequence,
+    fetch_darwin_dataset,
+    load_pil_image,
+    polygon_area,
+)
+
+
+def get_dataset(
+    dataset_name: str,
+    image_set: Optional[str] = "train",
+    mode: Optional[str] = "raw",
+    transforms: Optional = None,
+    poly_to_mask: Optional[bool] = False,
+    client: Optional[Client] = None,
+    **kwargs,
+):
+    """
+    Pulls a dataset from Darwin and returns a Dataset class that can be used with a PyTorch dataloader
+
+    Parameters
+    ----------
+    dataset_name : TODO type
+        Identifier of the dataset in Darwin
+    image_set :  TODO type
+        Split set [train, val, test]
+    mode :  TODO type
+        selects the dataset type [image_classification, instance_segmentation, semantic_segmentation]
+    transforms :  TODO type
+        List of PyTorch transforms
+    poly_to_mask :
+         TODO description
+    client:  TODO type
+        Darwin's client
+
+    TODO these entries are not part of the parameters (in the signature) and do not exist below: remove or fix?
+    val_percentage:  TODO type
+        percentage of images used in the validation set
+    test_percentage:  TODO type
+        percentage of images used in the validation set
+    force_fetching:  TODO type
+        discard local dataset and pull again from Darwin
+    force_resplit:  TODO type
+        discard previous split and create a new one
+    split_seed:  TODO type
+        fix seed for random split creation
+
+    Returns
+    -------
+    dataset : Dataset
+        Dataset class
+    """
+
+    root, split_id = fetch_darwin_dataset(dataset_name, client, **kwargs)
+
+    if mode == "raw":
+        dataset = Dataset(root, image_set=image_set, split_id=split_id, transforms=transforms)
+    elif mode == "classification":
+        dataset = ClassificationDataset(
+            root, image_set=image_set, split_id=split_id, transforms=transforms
+        )
+    elif mode == "instance_segmentation":
+        import darwin.torch.transforms as T
+
+        trfs = [T.ConvertPolysToInstanceMasks()]
+        if transforms is not None:
+            trfs.append(transforms)
+        transforms = T.Compose(trfs)
+        dataset = InstanceSegmentationDataset(
+            root, image_set=image_set, split_id=split_id, transforms=transforms
+        )
+    elif mode == "semantic_segmentation":
+        import darwin.torch.transforms as T
+
+        trfs = [T.ConvertPolysToMask()]
+        if transforms is not None:
+            trfs.append(transforms)
+        transforms = T.Compose(trfs)
+        dataset = SemanticSegmentationDataset(
+            root, image_set=image_set, split_id=split_id, transforms=transforms
+        )
+    else:
+        raise ValueError("Dataset type {mode} not supported.")
+
+    return dataset
+
+
+class Dataset(object):
+    def __init__(self, root: Path, image_set: str, split_id: Optional[str] = None, transforms=None):
```

Comment:
Genuine question, what is the different between: ` split_id: Optional[str] = None` and `transforms=None` in terms of type notation? I mean, `transforms` is also optional from the Python POV as there is a default argument so why would we go using the Optional[type] notation? Follow up question: shouldn't we be consistent and use the same notation for all parameters? I am sorry if its a stupid question I never used this type specifications before.